rfcs/0116-artifact-service.md

```diff
@@ -0,0 +1,196 @@
+* We want better Artifact handling without increasing complexity of the Queue
+  codebase
+* We want to have more clear responsibilities for our services
+* We want to consolidate the handling of Artifacts into a single service
+* We want to build generalised software when possible as participants of a free
+  software community
+
+# Details
+
+* The Artifact service will start from the Queue's `artifacts.js` file
+* Auth might not be done with scopes and tc-auth, but a shared secret between
+  Queue and Artifacts
+* The Queue will block new tasks after task resolution by not issuing new
+  signed urls
+* Data storage in Artifacts will be done with Postgres and not Azure Table
+  Storage
+* We do not intend to have an operation to list artifacts in the API
+* The redirecting logic of the Artifact service will adhere to the HTTP spec
+  using 300-series redirects where possible
+* Where content-negotiation is impossible, the service will block creation of
+  artifacts when incoming requests aren't compatible with required response
+* We will support uploading artifacts to different regions so that we can
+
+## API
+This API is up for debate and is only an illustrative starting point.  
+
+### Generalised request format
+There will be a concept of a generalised HTTP request.  This will be an object shaped like
+```javascript
+{
+  method: 'GET',
+  url: 'https://www.example.com',
+  headers: {
+    'user-agent': 'taskcluster',
+  }
+}
+```
+
+* The body will not be included, but can be described through HTTP headers in the request
+* Duplicated HTTP headers will be unsupported (i.e. key -> string not key -> list of strings)
+* All query string, path, fragements are contained in a fully formed `url` property
+
+### Creation
+
+```
+PUT /artifacts/:name
+```
+
+This endpoint will take an origin and name.  It will return a `200` with a
+document which contains a list of generalised requests which could be run to
+complete the actual upload.  Please see the queue's current `blob` storage type
+for an example of this pattern.  The body of the request will contain metadata
+about the artifact in the shape:
+
+```javascript
+{
+  origin: '10.10.10.10' || 's3_us-west-2',
+  contentType: 'application/json',
+  contentLength: 1234,
+  contentSha256: 'abcdef1234',
+  transferLength: 1234,
+  transferSha256: 'abcdef1234',
+  contentEncdoing: 
+  expiration: new Date(),
+}
+```
+
+The `origin` parameter will either be an IPv4 address, IPv6 address, or an
+identifier.  This parameter will specify the source location of the ultimately
+originating request.  This means the origin of the request to the Queue from
+the worker and not the request to the artifact service.  If the origin is an
+IPv4 or IPv6, a mapping of the address and IP will occur to find the backing
+storage.  If the origin is not an IPv4 or IPv6 address and is identical to a
+set of known and mappable identifiers (e.g. `s3_us-west-2`), then that will be
```

Comment:
I'm not sure we should allow identifiers, other than IPs, but I'm not sure...

Reply:
In the general case, I think this could be useful.  For example, when glandium wants to upload something from a coffeeshop in Nagoya, he should be able to specify 'ap-northeast-1` to get a quick upload, without requiring us to configure the IP range of that coffeeshop in the service (but requiring that we configure that region)

Reply:
yes, that's the idea.  I think in the general case, this is a critical feature and not one that's really all that difficult to implement.  We can verify that the configured identifiers aren't valid IP addresses, which makes this safe to do.  This would also support things like having the queue use a specific region or service for its unit tests.  It would also allow something like saying "All private artifacts must go to s3-us-west-2 for auditing reasons" or any of a bunch of other possible rules based on payload.

That said, I'm sure we could use some sort of geo-ip database service to do 'smart' things.  I think using a Geo-ip service for this would be an extreme amount of overkill.

Until there's a specific concern which cannot be addressed, let's stick with these identifiers.

Reply:
The counter argument against allowing this is that the service is supposed to abstract the fact that the storage is geographically distributed. Hence, we shouldn't allow people to peak past the abstraction.

It's a nit :)

Besides glandium will upload through the queue which will just take the IP of the caller as `origin`.

Reply:
The assumption is that ip blocks are published for every single cloud provider and that any possible user of this service has ip blocks of all of their users.  I don't think we can make the latter assumption.

It's also super annoying as a person debugging artifacts to have to launch an EC2 instance in eu-central-1 to be able to get that artifact's copy in eu-central-1.

Reply:
`origin` is just parameter, you can make up and IP for debugging...


I don't have strong objections... just saying we could reduce the surface exposed here.

Reply:
If we add/remove origins, then what do we do to strings that are no-longer valid? treat them as public internet, like IPs for which we have no cache. I guess that could work.

But it means that `origin` will be a free form string field, and not a ip-pattern or choice from enum.

Reply:
My thinking was that if it is not an IPv4/v6 and not a *known* id, that it would generate an error, so that misconfigured usage is immediately apparent.